### PR TITLE
Feature/gate of knowledge wishimpl

### PR DESCRIPTION
**Pull request by Phobossus**

### CODE REVIEW

File: RP-Core/src/main/java/abilitiesV2/skills/abitilies/knowledge/gateofknowledge/emissary/KnowledgeEmissary.java [hunk 1]
```diff
@@ -26,15 +31,59 @@ public KnowledgeEmissary(Location spawnLocation, Disguise appearance, Personalit
         this.emissaryMessageLists = EmissaryConfig.getDefault(personality);
         this.appearance = appearance;
         this.personality = personality;
+        this.emissaryWishlessBehaviour = new EmissaryWishlessBehaviour();
         this.setAppearance();
     }
 
+    public void setConversable(boolean conversable) {
+        this.conversable = conversable;
+    }
```
Comment (Pedrokase) [inline RP-Core/src/main/java/abilitiesV2/skills/abitilies/knowledge/gateofknowledge/emissary/KnowledgeEmissary.java:40]: Nitpick, but why not use `@Setter` ?

File: RP-Core/src/main/java/abilitiesV2/skills/abitilies/knowledge/gateofknowledge/emissary/KnowledgeEmissary.java [hunk 2]
```diff
@@ -26,15 +31,59 @@ public KnowledgeEmissary(Location spawnLocation, Disguise appearance, Personalit
...
+
+    public boolean heardWish() {
+        return heardWish;
+    }
+
+    public void setHeardWish(boolean heardWish) {
+        this.heardWish = heardWish;
+    }
```
Comment (Pedrokase) [inline RP-Core/src/main/java/abilitiesV2/skills/abitilies/knowledge/gateofknowledge/emissary/KnowledgeEmissary.java:48]: Same here: use `@Setter`

File: RP-Core/src/main/java/abilitiesV2/skills/abitilies/knowledge/gateofknowledge/skill/GateOfKnowledge.java [hunk 3]
```diff
@@ -46,30 +61,70 @@ public SkillDefinition getBaseSkillDefinition() {
...
                 .type(VariableEffect.Type.HAS_GATE_OF_KNOWLEDGE)
-                .duration(DURATION_SECONDS)
+                .durationAndActions(DURATION_SECONDS, ImmutableMap.of(
+                        welcomeMessageSecond, () -> {
+                            knowledgeEmissary.speak(EmissaryMessageType.WELCOME);
+                        },
+                        readyToHearWishesSecond, () -> {
+                            knowledgeEmissary.speak("takes a look at you, awaiting your response.",
+                                    RpChatMessageMode.EMOTE);
+                            knowledgeEmissary.setConversable(true);
```
Comment (Pedrokase) [inline RP-Core/src/main/java/abilitiesV2/skills/abitilies/knowledge/gateofknowledge/skill/GateOfKnowledge.java:95]: Won't the emote display to everyone? So 10 players nearby could hear "X takes a look at you..."

File: RP-Core/src/main/java/abilitiesV2/skills/abitilies/knowledge/gateofknowledge/wishes/EscapeAllTroublesWish.java [hunk 4]
```diff
@@ -1,38 +1,113 @@
...
                     @Override
                     public void acceptWish(LivingCharacter caster, KnowledgeEmissary emissary) {
+                        emissary.speak(EmissaryMessageType.WISH_ACCEPTED, GateOfKnowledge.WISH_GRANT_RESPONSE_DELAY_SECONDS);
+                        runWithDelay(caster, () -> {
+                            Location tpLocation = LocationUtils.getRandom(caster.getValidLivingEntity().getWorld());
 
+                            playTpAnimation(caster);
+                            caster.getValidLivingEntity().teleport(tpLocation);
+                            playTpAnimation(caster);
+                        }, GateOfKnowledge.WISH_GRANT_ACTION_DELAY_SECONDS);
```
Comment (Pedrokase) [inline RP-Core/src/main/java/abilitiesV2/skills/abitilies/knowledge/gateofknowledge/wishes/EscapeAllTroublesWish.java:49]: Seeing this implementation, won't we have too much repetitive code for each Wish? 
I'm meaning the
```java
 emissary.speak(EmissaryMessageType.WISH_ACCEPTED,GateOfKnowledge.WISH_GRANT_RESPONSE_DELAY_SECONDS);
  runWithDelay(caster, () -> {...})
```

Comment (Phobossus) [reply RP-Core/src/main/java/abilitiesV2/skills/abitilies/knowledge/gateofknowledge/wishes/EscapeAllTroublesWish.java:49]: My assumption is that certain wishes might choose to display something else than WISH_ACCEPTED, or divide acceptation into more than 1 kind (with respective messages and other actions), hence emissary.speak(ACCEPTED) isn't called automatically.

runWithDelay - that depends on whether every wish is supposed to be executed after exact same time, or perhaps some will be instant and some not.

**Review by Pedrokase — APPROVED** [review]

**Pull request merged**